compilation of master branch fails in alpine (docker)

Observed the following error today (14.8), but same process worked yesterday (13.8.):

`g++ -c -DDMLC_LOG_CUSTOMIZE=1 -std=c++11 -Wall -Wno-unknown-pragmas -Iinclude   -Idmlc-core/include -Irabit/include -I/include -O3 -funroll-loops -msse2 -fPIC -fopenmp src/predictor/cpu_predictor.cc -o build/predictor/cpu_predictor.o
In file included from src/common/./host_device_vector.h:14:0,
                 from src/common/host_device_vector.cc:11:
src/common/./span.h:437:22: error: 'ptrdiff_t' has not been declared
   template <class U, ptrdiff_t OtherExtent,
src/common/./span.h:440:49: error: 'OtherExtent' was not declared in this scope
               detail::IsAllowedExtentConversion<OtherExtent, Extent>::value>>
                                                 ^~~~~~~~~~~
src/common/./span.h:440:68: error: template argument 1 is invalid
               detail::IsAllowedExtentConversion<OtherExtent, Extent>::value>>
                                                                    ^
src/common/./span.h:441:47: error: 'OtherExtent' was not declared in this scope
   XGBOOST_DEVICE constexpr Span(const Span<U, OtherExtent>& _other)   // NOLINT
                                               ^~~~~~~~~~~
src/common/./span.h:441:58: error: template argument 2 is invalid
   XGBOOST_DEVICE constexpr Span(const Span<U, OtherExtent>& _other)   // NOLINT
                                                          ^
src/common/./span.h: In constructor 'constexpr xgboost::common::Span<T, Extent>::Span(const int&)':
src/common/./span.h:442:38: error: request for member 'size' in '_other', which is of non-class type 'const int'
       __span_noexcept : size_(_other.size()), data_(_other.data()) {}
                                      ^~~~
src/common/./span.h:442:60: error: request for member 'data' in '_other', which is of non-class type 'const int'
       __span_noexcept : size_(_other.size()), data_(_other.data()) {}
`

the important part of the build (from dockerfile)
` git clone --recursive https://github.com/dmlc/xgboost && \
    sed -i '/#define DMLC_LOG_STACK_TRACE 1/d' /src/xgboost/dmlc-core/include/dmlc/base.h && \
    sed -i '/#define DMLC_LOG_STACK_TRACE 1/d' /src/xgboost/rabit/include/dmlc/base.h && \
    ln -s locale.h /usr/include/xlocale.h && \
    cd /src/xgboost &&  make -j4`

(stolen  from this [Dockerfile](https://github.com/Nexosis/alpine-xgboost/blob/master/Dockerfile), referred from StackOverflow (sry, no link),   referred from [3157](https://github.com/dmlc/xgboost/issues/3157) ...)

i think the commit [2c502784fffeb4aa35734d70ede47c98a8639fc2](https://github.com/dmlc/xgboost/commit/2c502784fffeb4aa35734d70ede47c98a8639fc2) is the culprit, as it add an  include for the span.h mentioned in the error message.

Reverting is an option, 
but what's the reason for that include and 
how does this break the build on alpine/musl?

--- a/src/common/span.h
+++ b/src/common/span.h
@@ -31,8 +31,8 @@
 
 #include <xgboost/logging.h>  // CHECK
 
-#include <cinttypes>    // int64_t
-#include <type_traits>  // remove_cv_t
+#include <cinttypes>          // int64_t
+#include <type_traits>
 
 /*!
  * The version number 1910 is picked up from GSL.
@@ -237,9 +237,9 @@ class SpanIterator {
 /*!
  * The extent E of the span returned by subspan is determined as follows:
  *
- *   - If Count is not std::dynamic_extent, Count;
- *   - Otherwise, if Extent is not std::dynamic_extent, Extent - Offset;
- *   - Otherwise, std::dynamic_extent.
+ *   - If Count is not dynamic_extent, Count;
+ *   - Otherwise, if Extent is not dynamic_extent, Extent - Offset;
+ *   - Otherwise, dynamic_extent.
  */
 template <detail::ptrdiff_t Extent,
           detail::ptrdiff_t Offset,
@@ -269,7 +269,7 @@ struct IsAllowedElementTypeConversion : public std::integral_constant<
 template <class T>
 struct IsSpanOracle : std::false_type {};
 
-template <class T, std::ptrdiff_t Extent>
+template <class T, detail::ptrdiff_t Extent>
 struct IsSpanOracle<Span<T, Extent>> : std::true_type {};
 
 template <class T>
@@ -434,7 +434,7 @@ class Span {
   XGBOOST_DEVICE Span(const Container& _cont) : size_(_cont.size()),  // NOLINT
                                                 data_(_cont.data()) {}
 
-  template <class U, ptrdiff_t OtherExtent,
+  template <class U, detail::ptrdiff_t OtherExtent,
             class = typename std::enable_if<
               detail::IsAllowedElementTypeConversion<U, T>::value &&
               detail::IsAllowedExtentConversion<OtherExtent, Extent>::value>>
@@ -456,7 +456,7 @@ class Span {
     return {this, 0};
   }
 
-  XGBOOST_DEVICE constexpr iterator end() const __span_noexcept {   // NOLINT
+  XGBOOST_DEVICE constexpr iterator end() const __span_noexcept {    // NOLINT
     return {this, size()};
   }
 
@@ -464,7 +464,7 @@ class Span {
     return {this, 0};
   }
 
-  XGBOOST_DEVICE constexpr const_iterator cend() const __span_noexcept {   // NOLINT
+  XGBOOST_DEVICE constexpr const_iterator cend() const __span_noexcept {    // NOLINT
     return {this, size()};
   }
 
@@ -538,8 +538,8 @@ class Span {
    * If Count is std::dynamic_extent, r.size() == this->size() - Offset;
    * Otherwise r.size() == Count.
    */
-  template < detail::ptrdiff_t Offset,
-             detail::ptrdiff_t Count = dynamic_extent >
+  template <detail::ptrdiff_t Offset,
+            detail::ptrdiff_t Count = dynamic_extent>
   XGBOOST_DEVICE auto subspan() const ->                   // NOLINT
       Span<element_type,
            detail::ExtentValue<Extent, Offset, Count>::value> {

